### ISSUE TITLE
Cannot upload new blob to existing container with SAS token generated by API for client

### ISSUE THREAD

**Issue by wojwarych** [Azure/Azurite#1551]
### Which service(blob, file, queue, table) does this issue concern?
blob
### Which version of the Azurite was used?
3.17.1
`pip list` for api:
```
Package                           Version
--------------------------------- ---------
azure-core                        1.24.1
azure-storage-blob                12.12.0
```
`pip list` for client:
```
Package                           Version
--------------------------------- ---------
azure-core                        1.24.1
azure-storage-blob                12.12.0
```
Both run under python `3.7.x`

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
DockerHub
### What problem was encountered?
`SAS` token that's generated by API server for client seems to be unsufficient so requests to `PUT` blobs to container directly via client app return `403`

### Steps to reproduce the issue?
Server API and Azurite living under same, docker-compose environment:
```yaml
version: "3.8"

services:
  web:
    build:
      context: ..
      dockerfile: ./docker/Dockerfile
    command: /start-reload.sh
    ports:
      - "8000:8000"
    volumes:
      - ../src:/app/src
      - ../tests/:/app/tests/
      - ../scripts/:/app/scripts/
    depends_on:
      - blob

  blob:
    image: mcr.microsoft.com/azure-storage/azurite:3.17.1
    ports:
      - "8888:8888"
      - "10000:10000"
    command: azurite-blob --blobHost 0.0.0.0 -d /data/debug.log
    volumes:
      - ../azurite:/data
 ```

To better understand flow:
1. API generates SAS token for client by this function:
```py
sas_token = generate_container_sas(
   self._container_client.account_name,  # devstoreaccount1
   self._container_client.container_name,  # name of container to which client has access to
   account_key=self._container_client.credential.account_key,  # Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==
   permissions=ContainerSasPermissions(write=True, create=True),
   expiry=datetime.utcnow() + timedelta(minutes=30),
   start=datetime.utcnow(),
)
return sas_token
```
Then client instantiates owns `BlobServiceClient` and tries to upload data to Azurite blob:
```py
from azure.storage.blob.aio import BlobServiceClient


class BlobClient:
    def __init__(self, url: str, sas_token: str):
        self._blob_service_client = BlobServiceClient("http://localhost:1000/devstoreaccount1", credential=sas_token) 

    async def upload(self, sample_id: int, analysis: Analysis):
        async with self._blob_service_client:
            container = self._blob_service_client.get_container_client(container_name)

            # returns a list of tuples where second item is tuple with name of blob and byte data
            media_data = MediaFiles.parse_dto_model(sample_id, analysis)  
            for file in media_data:
                blob_client = await container.upload_blob(
                    name=file[1][0], data=file[1][1]
                )

                props = await blob_client.get_blob_properties()
            return props
```
If possible, please provide the debug log using the -d parameter, replacing \<pathtodebuglog\> with an appropriate path for your OS, or review the instructions for docker containers:

```
2022-06-17T14:08:26.946Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: BlobStorageContextMiddleware: RequestMethod=PUT RequestURL=http://localhost/devstoreaccount1/testtest/QWERTYUIOPASDFGH/1995_09_15_22_57_21/MAST/parent.avi?st=2022-06-17T14:08:03Z&se=2022-06-17T14:38:03Z&spr=http&sv=2021-06-08&sr=c&sig=KXPXMUAOADXaBY7HS4EIIoqVN3QoSyHrT0Ti1%2BQ2IAI%3D RequestHeaders:{"host":"localhost:10000","content-length":"15","x-ms-blob-type":"BlockBlob","if-none-match":"*","x-ms-version":"2021-06-08","content-type":"application/octet-stream","accept":"application/xml","x-ms-date":"Fri, 17 Jun 2022 14:08:26 GMT","x-ms-client-request-id":"f5216948-ee46-11ec-9871-b40ede4c48ce","user-agent":"azsdk-python-storage-blob/12.12.0 Python/3.7.6 (Windows-10-10.0.19041-SP0)","accept-encoding":"gzip, deflate"} ClientIP=172.21.0.1 Protocol=http HTTPVersion=1.1
2022-06-17T14:08:26.946Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: BlobStorageContextMiddleware: Account=devstoreaccount1 Container=testtest Blob=QWERTYUIOPASDFGH/1995_09_15_22_57_21/MAST/parent.avi
2022-06-17T14:08:26.946Z dfc5a46c-3ff7-4e25-a157-577d715513b1 verbose: DispatchMiddleware: Dispatching request...
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: DispatchMiddleware: Operation=BlockBlob_Upload
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 verbose: AuthenticationMiddlewareFactory:createAuthenticationMiddleware() Validating authentications.
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: PublicAccessAuthenticator:validate() Start validation against public access.
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: PublicAccessAuthenticator:validate() Getting account properties...
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: PublicAccessAuthenticator:validate() Retrieved account name from context: devstoreaccount1, container: testtest, blob: QWERTYUIOPASDFGH/1995_09_15_22_57_21/MAST/parent.avi
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: PublicAccessAuthenticator:validate() Skip public access authentication. Cannot get public access type for container testtest
2022-06-17T14:08:26.948Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: BlobSharedKeyAuthenticator:validate() Start validation against account shared key authentication.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: BlobSharedKeyAuthenticator:validate() Request doesn't include valid authentication header. Skip shared key authentication.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: AccountSASAuthenticator:validate() Start validation against account Shared Access Signature pattern.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: AccountSASAuthenticator:validate() Getting account properties...
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: AccountSASAuthenticator:validate() Retrieved account name from context: devstoreaccount1, container: testtest, blob: QWERTYUIOPASDFGH/1995_09_15_22_57_21/MAST/parent.avi
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: AccountSASAuthenticator:validate() Got account properties successfully.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: AccountSASAuthenticator:validate() Retrieved signature from URL parameter sig: KXPXMUAOADXaBY7HS4EIIoqVN3QoSyHrT0Ti1+Q2IAI=
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: AccountSASAuthenticator:validate() Failed to get valid account SAS values from request.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: BlobSASAuthenticator:validate() Start validation against blob service Shared Access Signature pattern.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: BlobSASAuthenticator:validate() Getting account properties...
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: BlobSASAuthenticator:validate() Retrieved account name from context: devstoreaccount1, container: testtest, blob: QWERTYUIOPASDFGH/1995_09_15_22_57_21/MAST/parent.avi
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: BlobSASAuthenticator:validate() Got account properties successfully.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: BlobSASAuthenticator:validate() Retrieved signature from URL parameter sig: KXPXMUAOADXaBY7HS4EIIoqVN3QoSyHrT0Ti1+Q2IAI=
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 debug: BlobSASAuthenticator:validate() Signed resource type is c.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 warn: BlobSASAuthenticator:generateBlobSASSignature(): Must provide 'permissions' and 'expiryTime' for Blob SAS generation when 'identifier' is not provided.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: BlobSASAuthenticator:validate() Failed to get valid blob service SAS values from request. Skip blob service SAS validation.
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Received a MiddlewareError, fill error information to HTTP response
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: ErrorName=StorageError ErrorMessage=Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.  ErrorHTTPStatusCode=403 ErrorHTTPStatusMessage=Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature. ErrorHTTPHeaders={"x-ms-error-code":"AuthorizationFailure","x-ms-request-id":"dfc5a46c-3ff7-4e25-a157-577d715513b1"} ErrorHTTPBody="<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<Error>\n  <Code>AuthorizationFailure</Code>\n  <Message>Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.\nRequestId:dfc5a46c-3ff7-4e25-a157-577d715513b1\nTime:2022-06-17T14:08:26.949Z</Message>\n</Error>" ErrorStack="StorageError: Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.\n    at Function.getAuthorizationFailure (/opt/azurite/dist/src/blob/errors/StorageErrorFactory.js:134:16)\n    at /opt/azurite/dist/src/blob/middlewares/AuthenticationMiddlewareFactory.js:22:56\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
2022-06-17T14:08:26.949Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Set HTTP code: 403
2022-06-17T14:08:26.950Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Set HTTP status message: Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.
2022-06-17T14:08:26.950Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Set HTTP Header: x-ms-error-code=AuthorizationFailure
2022-06-17T14:08:26.950Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Set HTTP Header: x-ms-request-id=dfc5a46c-3ff7-4e25-a157-577d715513b1
2022-06-17T14:08:26.950Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Set content type: application/xml
2022-06-17T14:08:26.950Z dfc5a46c-3ff7-4e25-a157-577d715513b1 error: ErrorMiddleware: Set HTTP body: "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n<Error>\n  <Code>AuthorizationFailure</Code>\n  <Message>Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.\nRequestId:dfc5a46c-3ff7-4e25-a157-577d715513b1\nTime:2022-06-17T14:08:26.949Z</Message>\n</Error>"
2022-06-17T14:08:26.950Z dfc5a46c-3ff7-4e25-a157-577d715513b1 info: EndMiddleware: End response. TotalTimeInMS=4 StatusCode=403 StatusMessage=Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature. Headers={"server":"Azurite-Blob/3.17.1","x-ms-error-code":"AuthorizationFailure","x-ms-request-id":"dfc5a46c-3ff7-4e25-a157-577d715513b1","content-type":"application/xml"}
```

Response that client receives from azurite:
```
E   azure.core.exceptions.HttpResponseError: Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.
E   RequestId:dfc5a46c-3ff7-4e25-a157-577d715513b1
E   Time:2022-06-17T14:08:26.949Z
E   ErrorCode:AuthorizationFailure
E   Content: <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
E   <Error>
E     <Code>AuthorizationFailure</Code>
E     <Message>Server failed to authenticate the request. Make sure the value of the Authorization header is formed correctly including the signature.
E   RequestId:dfc5a46c-3ff7-4e25-a157-577d715513b1
E   Time:2022-06-17T14:08:26.949Z</Message>
E   </Error>
```

### Have you found a mitigation/solution?
I tried to upload data to real blob container on Azure and looked that data was uploaded successfully.  
I created manually a SAS token for specific container with `read` and `create` permissions, start time and expiry time as well as accepted only https connections. In the end, the blob data has appeared in container.  

Thanks for help in advance. Glad to answer any questions if you have.

**Comment by blueww** [comment]
@wojwarych 
From the Azurite debug log, we can see the error is : "Must provide 'permissions' and 'expiryTime' for Blob SAS generation when 'identifier' is not provided."
The request Uri in debug log is: `http://localhost/devstoreaccount1/testtest/QWERTYUIOPASDFGH/1995_09_15_22_57_21/MAST/parent.avi?st=2022-06-17T14:08:03Z&se=2022-06-17T14:38:03Z&spr=http&sv=2021-06-08&sr=c&sig=[hidden]`

There are no permission ("sp=") specific in this SAS, so it failed on SAS validation.
This looks not a valid SAS, please specifc proper permission in the SAS.

For your success upload on real blob container, please check if the SAS already has permission, or save access policy with include permission "si="..

**Comment by wojwarych** [comment]
@blueww thanks for the response  

Yeah, the SAS token on real blob has an `sp=` permission, while azurite does not. Which was weird for me, because I'm adding `ContainerSasPermissions` to `generate_container_sas` function.  
Anyways, I looked once again at that snippet and I see that I mistyped an argument. I passed `permissions` instead of `permission` there. I fixed that and all got worked. I'm closing this issue. :)
